### PR TITLE
Support for reading arrow RecordBatchReader

### PR DESCRIPTION
**Pull request by Arttii**
This is a very naive initial draft

### CODE REVIEW

File: tools/pythonpkg/src/arrow_array_stream.cpp [hunk 1]
```diff
@@ -35,20 +39,34 @@ unique_ptr<ArrowArrayStreamWrapper> PythonTableArrowArrayStreamFactory::Produce(
...
 		}
 
+	} else if (py_object_type == "RecordBatchReader") {
+		// py::object arrow_batch_scanner = py::module_::import("pyarrow.dataset").attr("Scanner").attr("from_batches");
+		// scanner = arrow_batch_scanner(arrow_obj_handle);
+		// auto record_batches = scanner.attr("to_reader")();
+		auto res = make_unique<ArrowArrayStreamWrapper>();
+		auto export_to_c = arrow_obj_handle.attr("_export_to_c");
+		export_to_c((uint64_t)&res->arrow_array_stream);
+		return res;
```
Comment (pdet) [inline tools/pythonpkg/src/arrow_array_stream.cpp:49]: I think you can just add ```|| py_object_type == "RecordBatchReader")``` to the if clause code above, the recordbatchreader should follow the same path as the table.

**Comment by pdet** [comment]
Can you also add it to the register function in py_connection?

**Comment by pdet** [comment]
and a test with the register function ofc

File: tools/pythonpkg/src/arrow_array_stream.cpp (hunk 1 above)
Comment (Arttii) [reply tools/pythonpkg/src/arrow_array_stream.cpp:49]: The arrow dataset does not seem to accept a reader as input. It says it should in the docs, but it's not implemented in the [code](https://github.com/apache/arrow/blob/master/python/pyarrow/dataset.py#L668) at least as far as I can see.

Comment (Arttii) [reply tools/pythonpkg/src/arrow_array_stream.cpp:49]: So the scanner that is created from_batches is per definition single use and we cannot make a dataset from the reader. Even if we pass scanner coming from a reader to the dataset it wont work as after any query the generator will be consumed. I guess one can create a view and then create an actual table to hack around this a bit, but that is not optimal. Or reread from the Flight Server.

Comment (pdet) [reply tools/pythonpkg/src/arrow_array_stream.cpp:49]: Hm, I see
I can have a look on it next week!

Comment (Arttii) [reply tools/pythonpkg/src/arrow_array_stream.cpp:49]: I discussed this a bit on the arrow issues [boards](https://github.com/apache/arrow/issues/12667) and this is the correct behaviour basically.

**Comment by pdet** [comment]
Hey, @Arttii I had a look at this, this morning, supporting it is a bit more involved than I initially thought. Since we can only produce one scanner on top of a record batch reader.

Basically, a scanner is created on the binder and per-thread (on the multithreaded case).

I think it might be possible to avoid this multi-scanner creation, I'll give it a go today.

**Comment by Arttii** [comment]
@pdet Thanks for taking a look! I thought that as well, but was not sure where to look at this. Let me know if I can help with anything.

**Comment by pdet** [comment]
Hey @Arttii many thanks again for giving this a go, I'm sorry I initially misread the difficulty level of this task.

I'm closing this PR in favor of #3299

**Pull request closed without merging**